### PR TITLE
Bug fix/fix suspicous stuff

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

I turned on various clang specific warnings, like "suspicous conversions".
To make it compile at all I added a new cmake option to turn of -Werror

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is a trivial rework / code cleanup without any test coverage.

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/8468/

### CODE REVIEW

File: arangod/Aql/RestAqlHandler.cpp [hunk 1]
```diff
@@ -98,7 +98,7 @@ std::pair<double, std::shared_ptr<VPackBuilder>> RestAqlHandler::getPatchedOptio
   {
     VPackObjectBuilder guard(options.get());
     TRI_ASSERT(optionsSlice.isObject());
-    for (auto const& pair : VPackObjectIterator(optionsSlice)) {
+    for (auto pair : VPackObjectIterator(optionsSlice)) {
```
Comment (dothebart) [inline arangod/Aql/RestAqlHandler.cpp]: any reason you didn't remove auto here too?

File: arangod/Aql/RestAqlHandler.cpp [hunk 2]
```diff
@@ -278,7 +278,7 @@ bool RestAqlHandler::registerSnippets(
   answerBuilder.openObject();
   // NOTE: We need to clean up all engines if we bail out during the following
   // loop
-  for (auto const& it : VPackObjectIterator(snippetsSlice)) {
+  for (auto it : VPackObjectIterator(snippetsSlice)) {
```
Comment (dothebart) [inline arangod/Aql/RestAqlHandler.cpp]: any reason you didn't remove auto here too?

File: arangod/CMakeLists.txt [hunk 3]
```diff
@@ -985,9 +985,9 @@ foreach(TARGET
 )
   target_include_directories(${TARGET} PUBLIC "${PROJECT_SOURCE_DIR}/${ENTERPRISE_INCLUDE_DIR}")
 
-  if(MSVC)
+  if(MSVC AND FAIL_ON_WARNINGS)
```
Comment (dothebart) [inline arangod/CMakeLists.txt]: please add a separate PR for this, we may want to have it easily backporteable.

File: arangod/Cluster/ClusterInfo.cpp [hunk 4]
```diff
@@ -1955,7 +1955,7 @@ Result ClusterInfo::createCollectionsCoordinator(
     }
 
     std::map<ShardID, std::vector<ServerID>> shardServers;
-    for (auto const& pair : VPackObjectIterator(info.json.get("shards"))) {
+    for (auto pair : VPackObjectIterator(info.json.get("shards"))) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterInfo.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ClusterMethods.cpp [hunk 5]
```diff
@@ -1216,7 +1216,7 @@ int selectivityEstimatesOnCoordinator(ClusterFeature& feature, std::string const
       }
 
       // add to the total
-      for (auto const& pair : VPackObjectIterator(answer.get("indexes"), true)) {
+      for (auto pair : VPackObjectIterator(answer.get("indexes"), true)) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterMethods.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ClusterMethods.cpp [hunk 6]
```diff
@@ -2025,7 +2025,7 @@ void fetchVerticesFromEngines(
     }
     
     bool cached = false;
-    for (auto const& pair : VPackObjectIterator(resSlice)) {
+    for (auto pair : VPackObjectIterator(resSlice)) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterMethods.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ClusterMethods.cpp [hunk 7]
```diff
@@ -2613,7 +2613,7 @@ arangodb::Result hotBackupList(std::vector<ServerID> const& dbServers, VPackSlic
       plan.add(resSlice.get("agency-dump")[0]);
     }
 
-    for (auto const& backup : VPackObjectIterator(resSlice.get("list"))) {
+    for (auto backup : VPackObjectIterator(resSlice.get("list"))) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterMethods.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ClusterMethods.cpp [hunk 8]
```diff
@@ -2711,7 +2711,7 @@ arangodb::Result matchBackupServersSlice(VPackSlice const planServers,
 
   // Skip all direct matching names in pair and remove them from localCopy
   std::unordered_set<std::string>::iterator it;
-  for (auto const& planned : VPackObjectIterator(planServers)) {
+  for (auto planned : VPackObjectIterator(planServers)) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterMethods.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ClusterMethods.cpp [hunk 9]
```diff
@@ -2888,13 +2888,13 @@ arangodb::Result applyDBServerMatchesToPlan(VPackSlice const plan,
                                                  std::map<ServerID, ServerID> const& matches) {
     if (s.isObject()) {
       VPackObjectBuilder o(&newPlan);
-      for (auto const& it : VPackObjectIterator(s)) {
+      for (auto it : VPackObjectIterator(s)) {
```
Comment (dothebart) [inline arangod/Cluster/ClusterMethods.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/FollowerInfo.cpp [hunk 10]
```diff
@@ -445,7 +445,7 @@ VPackBuilder FollowerInfo::newShardEntry(VPackSlice oldValue) const {
     VPackObjectBuilder b(&newValue);
     // Copy all but SERVERS and FailoverCandidates.
     // They will be injected later.
-    for (auto const& it : VPackObjectIterator(oldValue)) {
+    for (auto it : VPackObjectIterator(oldValue)) {
```
Comment (dothebart) [inline arangod/Cluster/FollowerInfo.cpp]: any reason you didn't remove auto here too?

File: arangod/Cluster/ServerState.cpp [hunk 11]
```diff
@@ -438,7 +438,7 @@ bool ServerState::integrateIntoCluster(ServerState::RoleEnum role,
     if (valueSlice.isObject()) {
       // map from server UUID to endpoint
       std::unordered_map<std::string, std::string> endpoints;
-      for (auto const& it : VPackObjectIterator(valueSlice)) {
+      for (auto it : VPackObjectIterator(valueSlice)) {
```
Comment (dothebart) [inline arangod/Cluster/ServerState.cpp]: any reason you didn't remove auto here too?

File: arangod/Network/Utils.cpp [hunk 12]
```diff
@@ -163,7 +163,7 @@ void errorCodesFromHeaders(network::Headers headers,
       return;
     }
 
-    for (auto const& code : VPackObjectIterator(codesSlice)) {
+    for (auto code : VPackObjectIterator(codesSlice)) {
```
Comment (dothebart) [inline arangod/Network/Utils.cpp]: any reason you didn't remove auto here too?

File: arangod/Pregel/Conductor.cpp [hunk 13]
```diff
@@ -579,7 +579,7 @@ int Conductor::_initializeWorkers(std::string const& suffix, VPackSlice addition
     b.add(Utils::lazyLoadingKey, VPackValue(_lazyLoading));
     b.add(Utils::useMemoryMaps, VPackValue(_useMemoryMaps));
     if (additional.isObject()) {
-      for (auto const& pair : VPackObjectIterator(additional)) {
+      for (auto pair : VPackObjectIterator(additional)) {
```
Comment (dothebart) [inline arangod/Pregel/Conductor.cpp]: any reason you didn't remove auto here too?

File: arangod/Pregel/WorkerConfig.cpp [hunk 14]
```diff
@@ -75,15 +75,15 @@ void WorkerConfig::updateConfig(VPackSlice params) {
   }
 
   // To access information based on a user defined collection name we need the
-  for (auto const& it : VPackObjectIterator(collectionPlanIdMap)) {
+  for (auto it : VPackObjectIterator(collectionPlanIdMap)) {
```
Comment (dothebart) [inline arangod/Pregel/WorkerConfig.cpp]: any reason you didn't remove auto here too?

File: lib/Zip/unzip.cpp [hunk 15]
```diff
@@ -1085,7 +1085,7 @@ unzGetCurrentFileInfo(unzFile file, unz_file_info* pfile_info, char* szFileName,
     pfile_info->internal_fa = file_info64.internal_fa;
     pfile_info->external_fa = file_info64.external_fa;
 
-    pfile_info->tmu_date = file_info64.tmu_date,
+    pfile_info->tmu_date = file_info64.tmu_date;
```
Comment (dothebart) [inline lib/Zip/unzip.cpp]: WTF. good spot.

File: arangod/Aql/RestAqlHandler.cpp (hunk 1 above)
Comment (graetzer) [reply arangod/Aql/RestAqlHandler.cpp]: VPackObjectIterator::ObjectPair is a tad long

**Pull request merged**